samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -611,14 +616,30 @@ private boolean isReady() {
       if (coordinatorRequests.commitRequests().remove(taskName)) {
         needCommit = true;
       }
-      if (needWindow || needCommit || endOfStream) {
-        // ready for window or commit only when no messages are in progress and
-        // no window/commit in flight
+
+      boolean windowOrCommitInFlight = windowInFlight || commitInFlight;
+      /*
+       * A task is ready to commit, when task.commit(needCommit) is requested either by user or commit thread
+       * and either of the following conditions are true.
+       * a) When process, window, commit are not in progress.
+       * b) When task.async.commit is true and window, commit are not in progress.
+       */
+      if (needCommit) {
+        return (messagesInFlight.get() == 0 || isAsyncCommitEnabled) && !windowOrCommitInFlight;
+      } else if (needWindow || endOfStream) {
+        /*
+         * A task is ready for window operation, when task.window(needWindow) is requested by either user or window thread
+         * and window, commit are not in progress.
+         */
         return messagesInFlight.get() == 0 && !windowOrCommitInFlight;
```

Comment:
Question: if we need messagesInFlight.get() == 0, that would mean that the commit will have to wait till all in-flight messages to be completed? Wouldn't it still cause bursts in message process rate? I thought that we need to allow commit and process can always be scheduled simultaneously in order to perform async commit? The current logic seems to allow process() to be scheduled when async commit is in-flight, but not the other way. Wouldn't that cause stopage in processing when a commit is waiting (i.e. needCommit == true)? I don't think that this would have significant reduction of burstiness, since the difference before and after this change is simply more process() can be scheduled when commit is in flight. However, the number of commits are low and the latency of a single commit is not large. The burstiness caused by stopping the commit and process while waiting the inflight process() to be completed is not reduced, which is the main cause of the burstiness.

I thought that the right approach is to:
1) allow process and commit to always be scheduled simultaneously
2) bookkeeping the low watermark of each input SSP's offsets while callbacks are triggered and only checkpoint if the low watermark proceeds in commit

Happy to discuss in person.